// grunt-nuborn/processor/ResourceProcessor.js
// allow inheritance
require("../common/inheritance");

// load grunt API
var grunt = require("grunt");

// load path API
var nodePath = require("path");

// system path separator
var sep = nodePath.sep;

/**
* Template for grunt-contrib-copy configuration.
* Will be filled with values later dynamically.
*/
var copy = {
	main: {
		files: []
	}
};

/**
* @class ResourceProcessor
* It's job is to process the "resources" configuration part.
*/
var ResourceProcessor = Object.subClass(
{
	/**
	* Copy static resources which match the fileset patterns to defined output.
	* @param targets the list of targets (android/ios/web...) 
	* @param fileset fileset configuration in gruntfile
	*/
	process: function(targets, fileset) 
	{
		if (!fileset) { 
			return;
		}

		var filesToCopy = copy.main.files;

		// for each entry in the fileset
		fileset.forEach(function(node, index, array) 
		{
			if (!node.patterns) {
				return;
			}

			// for each pattern
			for (var key in targets) 
			{
				if (targets[key] && node[key]) 
				{
					filesToCopy.push({
						expand: true,
						flatten: true,
						src: node.patterns,
						dest: "build" + sep + key + sep + node[key].output
					});
				}
			}
		}, this);

		// let's add some configuration for grunt-contrib-copy plugin
		grunt.config("copy", copy);

		// run it
		grunt.task.run("copy");
	}
});

module.exports = ResourceProcessor;
